### PR TITLE
Ptr array foreach improve by dcache prefetching

### PR DESCRIPTION
**Pull request by shuki-zanyovka**
## What
Further improve the ptr_array foreach performance by using the CPU d-cache prefetching.

## Why ?
Improve ptr_array foreach performance by ~10%.

## How ?
Using the CPU dcache prefetching feature.

### CODE REVIEW

File: src/ucs/sys/compiler_def.h [hunk 1]
```diff
@@ -135,7 +135,7 @@
  * @return Address of the container structure.
  */
 #define ucs_container_of(_ptr, _type, _member) \
-    ( (_type*)( (char*)(void*)(_ptr) - ucs_offsetof(_type, _member) )  )
+    ( (_type*)( (uintptr_t)(void*)(_ptr) - ucs_offsetof(_type, _member) )  )
```
Comment (yosefe) [inline src/ucs/sys/compiler_def.h:138]: why needed?

File: src/ucs/datastruct/ptr_array.h [hunk 2]
```diff
@@ -203,11 +203,12 @@ __ucs_ptr_array_for_each_get_step_size(ucs_ptr_array_t *ptr_array,
...
 
-    if (ucs_unlikely(__ucs_ptr_array_is_free((ucs_ptr_array_elem_t)elem))) {
-       size_elem = (elem >> UCS_PTR_ARRAY_FREE_AHEAD_SHIFT);
-    } else {
+    size_elem = (elem >> UCS_PTR_ARRAY_FREE_AHEAD_SHIFT);
+    if (ucs_likely(!__ucs_ptr_array_is_free((ucs_ptr_array_elem_t)elem))) {
        size_elem = 1;
     }
+    /* Prefetch the next item */
+    ucs_prefetch((const void *)&(ptr_array)->start[element_index + size_elem]);
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_array.h:211]: 1. excess ( ) around ptr_array
2. why need cast to const void*

File: src/ucs/sys/compiler_def.h (hunk 1 above)
Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h:138]: GCC10 fails to compile, a warning complaining on ptr[-2] indexing to an array.

Comment (yosefe) [reply src/ucs/sys/compiler_def.h:138]: so this fix is not really related to "Ptr array foreach improve by dcache prefetching"?

File: src/ucs/datastruct/ptr_array.h (hunk 2 above)
Comment (shuki-zanyovka) [reply src/ucs/datastruct/ptr_array.h:211]: 1. Will fix soon.
2. ucs_prefetch() is defined as follows,
src/ucs/sys/compiler_def.h:#define ucs_prefetch(p)            __builtin_prefetch(p)
Then, __builtin_prefetch() addr argument type is  (const void *)
__builtin_prefetch (const void * addr , ...).

File: src/ucs/sys/compiler_def.h (hunk 1 above)
Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h:138]: Not related to the same pull request, but I figured it can be part of the same PR...

Comment (alex--m) [reply src/ucs/sys/compiler_def.h:138]: To clarify: this also fixes a (new...) GCC v10 issue shuki found while testing the optimization.

BTW @shuki-zanyovka - according to the CI this is a problem on other GCCs... need to change this to something all these GCC-s will like :)

Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h:138]: Closing this PR, I will create 2 separate PRs for the 2 issues.

**Comment by shuki-zanyovka** [comment]
Closing this PR, will create 2 separate PRs for the 2 issues.

**Pull request closed without merging**